### PR TITLE
12826 Add Rack Type

### PR DESCRIPTION
**Pull request by arthanson**
### Fixes: #12826 

Add's a RackType and can set a Rack to use a Rack Type and on creation will get the fields from it. 

Rack fields associated with a rack_type are disabled on Rack form if rack_type is set.

If rack_type is set on a rack and fields associated with a rack_type are modified (via API) then a validation error is thrown.

### CODE REVIEW

**Comment by arthanson** [comment]
@jeremystretch updated this to disable rack fields on edit if rack_type is set, also raises error if any of the fields are modified (potentially from API).

File: netbox/dcim/models/racks.py [hunk 1]
```diff
@@ -29,9 +29,183 @@
...
+        target_field='name',
+        max_length=100,
+        blank=True
+    )
+    slug = models.SlugField(
+        verbose_name=_('slug'),
+        max_length=100,
+        unique=True
+    )
+    type = models.CharField(
```
Comment (jeremystretch) [inline netbox/dcim/models/racks.py:70]: IMO we should rename `type` to avoid confusion with RackType. Maybe `form_factor`, or something similar?

Comment (jeremystretch) [reply netbox/dcim/models/racks.py:70]: I've renamed it to `form_factor`.

File: netbox/dcim/models/racks.py [hunk 2]
```diff
@@ -271,6 +396,19 @@ def clean(self):
                     })
 
     def save(self, *args, **kwargs):
+        if (not self.pk) and self.rack_type:
```
Comment (arthanson) [inline netbox/dcim/models/racks.py:399]: I think will have to take out the (not self.pk) and just copy as if you edit an existing rack that did not have a rack-type and set a rack-type it doesn't get updated correctly - can see below I edited IDF128 and to have rack type same as track, but the weight is still the old one.
Racks | NetBox 2024-07-16 14-10-05

**Review by jeremystretch — APPROVED** [review]

**Pull request merged**